### PR TITLE
storage: Define the range lease duration as 3 times the raft election cycle to prevent any issues from occurring when the `--raft-tick-interval` flag is set.

### LINKED ISSUE #6815: storage: Ensure range lease duration is consistent with raft election timeouts
When a new range lease is granted, the new lease holder starts a raft election if it is not already leader. This makes the range unavailable for at least two network round trips, and potentially for a multiple of the raft election timeout if the election is contested (which I think is reasonably likely in the case of an election triggered by a lease transfer). If the `--raft-tick-interval` flag has been increased (which often but not necessarily indicates that network latency is high), the election may not conclude in time for the lease holder to renew its lease during the grace period, which makes it likely that the next lease holder will be a different node, beginning the cycle again.

The range lease duration should probably be increased in proportion to the raft tick interval. It should at least be higher than the tick interval, and probably higher than the election timeout - perhaps as high as 3x the election timeout (remember that the election timeout is randomized in the range [electionTimeout, 2*electionTimeout)), which would change it from 1s to 4.5s with our current defaults.

### PR DESCRIPTION
**Pull request by d4l3k**
Fixes #6815

---

This change is [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/7086)

### CODE REVIEW

**Comment by tamird** [comment]
Reviewed 5 of 5 files at r1.
Review status: all files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

*[storage/store.go, line 75 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7086#-KJgQe8thPEYnKn_RBZy:-KJgQe8u-RAemi_jcSfK:-323829707) ([raw file](https://github.com/cockroachdb/cockroach/blob/473bd2e181dbff4188b402a6319aa9573fb11f8f/storage/store.go#L75)):*
> ```Go
> 	// leaderLeaseRaftElectionMultiplier specifies what multiple the leader lease
> 	// active duration should be of the raft election timeout.
> 	leaderLeaseRaftElectionMultiplier = 3
> ```

`leaderLeaseRaftElectionTimeoutMultiplier`?

---

*[storage/store.go, line 405 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7086#-KJgQt4697xCNDoUb70H:-KJgQt4697xCNDoUb70I:313784977) ([raw file](https://github.com/cockroachdb/cockroach/blob/473bd2e181dbff4188b402a6319aa9573fb11f8f/storage/store.go#L405)):*
> ```Go
> 	// LeaderLeaseActiveDuration is the duration of the active period of leader
> 	// leases requested.
> 	LeaderLeaseActiveDuration time.Duration
> ```

how come this is exported?

---

*[storage/store.go, line 407 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7086#-KJgQwdBTwsRjUE3YY0n:-KJgQwdBTwsRjUE3YY0o:-787898301) ([raw file](https://github.com/cockroachdb/cockroach/blob/473bd2e181dbff4188b402a6319aa9573fb11f8f/storage/store.go#L407)):*
> ```Go
> 	LeaderLeaseActiveDuration time.Duration
> 
> 	// leaderLeaseRenewalDuration specifies a "time" interval at the
> ```

why is "time" quoted?

---

*[storage/store.go, line 411 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7086#-KJgQvF4arUC8g0bsvvh:-KJgQvF4arUC8g0bsvvi:313784977) ([raw file](https://github.com/cockroachdb/cockroach/blob/473bd2e181dbff4188b402a6319aa9573fb11f8f/storage/store.go#L411)):*
> ```Go
> 	// start of the stasis period) during which timestamps will trigger an
> 	// asynchronous renewal of the lease.
> 	LeaderLeaseRenewalDuration time.Duration
> ```

how come this is exported?

---

*[storage/store.go, line 647 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7086#-KJgR5Q91rfZCkefL93-:-KJgR5Q91rfZCkefL930:1531155169) ([raw file](https://github.com/cockroachdb/cockroach/blob/473bd2e181dbff4188b402a6319aa9573fb11f8f/storage/store.go#L647)):*
> ```Go
> 	raftElectionTimeout := time.Duration(sc.RaftElectionTimeoutTicks) * sc.RaftTickInterval
> 	sc.LeaderLeaseActiveDuration = leaderLeaseRaftElectionMultiplier * raftElectionTimeout
> 	sc.LeaderLeaseRenewalDuration = sc.LeaderLeaseActiveDuration / 5
> ```

how come the `5` here isn't a named constant?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7086)*

**Comment by d4l3k** [comment]
Review status: all files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

*[storage/store.go, line 75 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7086#-KJgQe8thPEYnKn_RBZy:-KJgRthWa7ICL8ejpb1u:1320461631) ([raw file](https://github.com/cockroachdb/cockroach/blob/473bd2e181dbff4188b402a6319aa9573fb11f8f/storage/store.go#L75)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> `leaderLeaseRaftElectionTimeoutMultiplier`?
</details>
Thought it was too wordy already. Done.

---

*[storage/store.go, line 405 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7086#-KJgQt4697xCNDoUb70H:-KJgSNznPwUnZs9ely7A:-2118672989) ([raw file](https://github.com/cockroachdb/cockroach/blob/473bd2e181dbff4188b402a6319aa9573fb11f8f/storage/store.go#L405)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> how come this is exported?
</details>
It was previously exported, and all other StoreContext parameters are exported. Done.

---

*[storage/store.go, line 407 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7086#-KJgQwdBTwsRjUE3YY0n:-KJgSXPA6EfqXpcTUs9w:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/473bd2e181dbff4188b402a6319aa9573fb11f8f/storage/store.go#L407)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> why is "time" quoted?
</details>
Done.

---

*[storage/store.go, line 411 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7086#-KJgQvF4arUC8g0bsvvh:-KJgSYfv4N4sEf25QpDJ:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/473bd2e181dbff4188b402a6319aa9573fb11f8f/storage/store.go#L411)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> how come this is exported?
</details>
Done.

---

*[storage/store.go, line 647 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7086#-KJgR5Q91rfZCkefL93-:-KJgSg2Dxa2lsMGf29Lv:-1662746470) ([raw file](https://github.com/cockroachdb/cockroach/blob/473bd2e181dbff4188b402a6319aa9573fb11f8f/storage/store.go#L647)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> how come the `5` here isn't a named constant?
</details>
It was previously an unnamed constant, and I'm not really sure what rational it had. Extracted into `leaderLeaseRenewalDivisor`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7086)*

**Comment by tamird** [comment]
:lgtm:

<details><summary><i>Previously, d4l3k (Tristan Rice) wrote…</i></summary>
> #### storage: Define the range lease duration as 3 times the raft election cycle to prevent any issues from occurring when the \`\-\-raft\-tick\-interval\` flag is set\.
> 
> Fixes #6815
> 
> 
> 
> 
> 
> 
</details>


---

Reviewed 3 of 3 files at r2.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---

*[storage/store.go, line 77 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7086#-KJgXzUmBNM9127qlBNh:-KJgXzUnaFfcgf0lQsAc:-2019604090) ([raw file](https://github.com/cockroachdb/cockroach/blob/901556824c0d74bae82ac5705cf3c004ef4d6792/storage/store.go#L77)):*
> ```Go
> 	leaderLeaseRaftElectionTimeoutMultiplier = 3
> 
> 	// leaderLeaseRenewalDivisor specifies what multiple the leader lease renewal
> ```

s/multiple/quotient/

---

*[storage/store.go, line 415 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7086#-KJgYJ1IbL-fVv-ZxhQQ:-KJgYJ1IbL-fVv-ZxhQR:-293502755) ([raw file](https://github.com/cockroachdb/cockroach/blob/901556824c0d74bae82ac5705cf3c004ef4d6792/storage/store.go#L415)):*
> ```Go
> 	// leaderLeaseRenewalDuration specifies a time interval at the end of the
> 	// active lease interval (i.e. bounded to the right by the start of the stasis
> 	// period) during which timestamps will trigger an asynchronous renewal of the
> ```

"timestamps will trigger" what does this mean? cc @tschottdorf who i think originally authored this comment

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7086#-:-KJgYO-ThkR9r7IQqLo8:1417088721)*

**Comment by d4l3k** [comment]
Review status: 4 of 5 files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---

*[storage/store.go, line 77 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7086#-KJgXzUmBNM9127qlBNh:-KJgZ2_qnP8Zv1zyO5Km:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/901556824c0d74bae82ac5705cf3c004ef4d6792/storage/store.go#L77)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> s/multiple/quotient/
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7086)*

**Comment by tschottdorf** [comment]
:lgtm:

<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> :lgtm:
</details>


---

Reviewed 5 of 5 files at r1, 3 of 3 files at r2, 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---

*[storage/store.go, line 415 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7086#-KJgYJ1IbL-fVv-ZxhQQ:-KJgYvzttKWkG0vGoQ0n:1056317426) ([raw file](https://github.com/cockroachdb/cockroach/blob/901556824c0d74bae82ac5705cf3c004ef4d6792/storage/store.go#L415)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> "timestamps will trigger" what does this mean? cc @tschottdorf who i think originally authored this comment
</details>
s/timestamps/operations/

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7086#-:-KJgZbs18AdrGWd-raSP:1417088721)*

**Comment by d4l3k** [comment]
Review status: 4 of 5 files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---

*[storage/store.go, line 415 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7086#-KJgYJ1IbL-fVv-ZxhQQ:-KJg_3oKVnZCzjOG2KbD:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/901556824c0d74bae82ac5705cf3c004ef4d6792/storage/store.go#L415)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> s/timestamps/operations/
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7086)*

**Comment by tamird** [comment]
Reviewed 1 of 1 files at r3, 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7086)*

**Comment by bdarnell** [comment]
LGTM, but note that this changes the default lease duration from 1s to 4.5s. I think that's probably a good idea given the underlying raft dynamics, but we should keep an eye on this and maybe revisit the numbers involved.

<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> :lgtm:
</details>


---

Reviewed 2 of 5 files at r1, 2 of 3 files at r2, 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7086#-:-KJhvE6THDAhGAClBFkV:-154394132)*

**Comment by tschottdorf** [comment]
Definitely also see how this affects the test runtime (incl acceptance)
seeing that most shutdown is now graceful and waits for lease expiration.

On Tue, Jun 7, 2016 at 11:08 PM Ben Darnell <notifications@github.com>
wrote:

> LGTM, but note that this changes the default lease duration from 1s to
> 4.5s. I think that's probably a good idea given the underlying raft
> dynamics, but we should keep an eye on this and maybe revisit the numbers
> involved.
>
> *Previously, tschottdorf (Tobias Schottdorf) wrote…*
>
> :lgtm:
>
> ------------------------------
>
> Reviewed 2 of 5 files at r1, 2 of 3 files at r2, 1 of 1 files at r4.
> Review status: all files reviewed at latest revision, 1 unresolved
> discussion, all commit checks successful.
> ------------------------------
>
> *Comments from Reviewable
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/7086#-:-KJhvE6THDAhGAClBFkV:-154394132>*
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/7086#issuecomment-224475835>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AE135H0wIvJ30Ue7tUba2jfoVxiSvdYgks5qJjIegaJpZM4IwL9W>
> .
>
-- 

-- Tobias

**Comment by d4l3k** [comment]
Here's some not particularly scientific tests that I ran on my laptop.
I ran `time make test` three times on each branch.

```
master (1s range lease)
  - make test  380.10s user 38.79s system 284% cpu 2:27.12 total
  - make test  362.97s user 36.63s system 297% cpu 2:14.26 total
  - make test  359.58s user 37.39s system 295% cpu 2:14.21 total

this PR (4.5s range lease)
  - make test  358.89s user 35.98s system 308% cpu 2:08.20 total
  - make test  375.23s user 38.30s system 287% cpu 2:23.89 total
  - make test  368.03s user 36.98s system 304% cpu 2:12.95 total
```

Not sure if there's any statistically significant difference.

**Comment by d4l3k** [comment]
make acceptance

```
master (1s range lease)
  - ok  	github.com/cockroachdb/cockroach/acceptance	147.374s
  - ok  	github.com/cockroachdb/cockroach/acceptance	157.884s
  - ok  	github.com/cockroachdb/cockroach/acceptance	156.997s

this PR (4.5s range lease)
  - ok  	github.com/cockroachdb/cockroach/acceptance	153.721s
  - ok  	github.com/cockroachdb/cockroach/acceptance	281.037s
  - ok  	github.com/cockroachdb/cockroach/acceptance	175.295s
```
